wit: `TestTypeDefRootOwnerNamesNotNil` failed for this WIT

Consider the following WIT:

```wit
package my:resources;

world resources {
  export exports: interface {
    resource x {
      constructor(a: f64);
      get-a: func() -> f64;
      set-a: func(a: f64);
      add: static func(x: x, a: f64) -> x;
    }
  }
}
```

The Resource `x` is a root TypeDef, but it doesn't have a owner name (the name is `nil`)

```
go test ./wit -update
--- FAIL: TestTypeDefRootOwnerNamesNotNil (0.10s)
    --- FAIL: TestTypeDefRootOwnerNamesNotNil/../testdata/codegen/resources.wit.json (0.00s)
        --- FAIL: TestTypeDefRootOwnerNamesNotNil/../testdata/codegen/resources.wit.json/TypeDefs[1]#x (0.00s)
            testdata_test.go:425: Owner.Name is nil
FAIL
FAIL    github.com/bytecodealliance/wasm-tools-go/wit   2.892s
FAIL
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/),
 - [#161](https://github.com/bytecodealliance/wasm-tools-go/issues/161): correctly handle `constructor` as a WIT keyword in wit.
 - [#165](https://github.com/bytecodealliance/wasm-tools-go/issues/165): fixed use of imported types in exported functions.
 - [#167](https://github.com/bytecodealliance/wasm-tools-go/issues/167): fixed a logic flaw in `TestHasBorrow`.
+- [#170](https://github.com/bytecodealliance/wasm-tools-go/issues/170): resolve implied names for interface imports and exports in a world.
 
 ## [v0.2.0] — 2024-09-05
 

--- a/testdata/issues/issue170.wit
+++ b/testdata/issues/issue170.wit
@@ -1,0 +1,12 @@
+package issues:issue170;
+
+world w {
+	export example: interface {
+		resource r {
+			constructor(a: f64);
+			get-a: func() -> f64;
+			set-a: func(a: f64);
+			add: static func(r: r, a: f64) -> r;
+		}
+	}
+}

--- a/testdata/issues/issue170.wit.json
+++ b/testdata/issues/issue170.wit.json
@@ -1,0 +1,135 @@
+{
+  "worlds": [
+    {
+      "name": "w",
+      "imports": {},
+      "exports": {
+        "example": {
+          "interface": {
+            "id": 0
+          }
+        }
+      },
+      "package": 0
+    }
+  ],
+  "interfaces": [
+    {
+      "name": null,
+      "types": {
+        "r": 0
+      },
+      "functions": {
+        "[constructor]r": {
+          "name": "[constructor]r",
+          "kind": {
+            "constructor": 0
+          },
+          "params": [
+            {
+              "name": "a",
+              "type": "f64"
+            }
+          ],
+          "results": [
+            {
+              "type": 2
+            }
+          ]
+        },
+        "[method]r.get-a": {
+          "name": "[method]r.get-a",
+          "kind": {
+            "method": 0
+          },
+          "params": [
+            {
+              "name": "self",
+              "type": 1
+            }
+          ],
+          "results": [
+            {
+              "type": "f64"
+            }
+          ]
+        },
+        "[method]r.set-a": {
+          "name": "[method]r.set-a",
+          "kind": {
+            "method": 0
+          },
+          "params": [
+            {
+              "name": "self",
+              "type": 1
+            },
+            {
+              "name": "a",
+              "type": "f64"
+            }
+          ],
+          "results": []
+        },
+        "[static]r.add": {
+          "name": "[static]r.add",
+          "kind": {
+            "static": 0
+          },
+          "params": [
+            {
+              "name": "r",
+              "type": 2
+            },
+            {
+              "name": "a",
+              "type": "f64"
+            }
+          ],
+          "results": [
+            {
+              "type": 2
+            }
+          ]
+        }
+      },
+      "package": 0
+    }
+  ],
+  "types": [
+    {
+      "name": "r",
+      "kind": "resource",
+      "owner": {
+        "interface": 0
+      }
+    },
+    {
+      "name": null,
+      "kind": {
+        "handle": {
+          "borrow": 0
+        }
+      },
+      "owner": null
+    },
+    {
+      "name": null,
+      "kind": {
+        "handle": {
+          "own": 0
+        }
+      },
+      "owner": null
+    }
+  ],
+  "packages": [
+    {
+      "name": "issues:issue170",
+      "interfaces": {},
+      "worlds": {
+        "w": 0
+      }
+    }
+  ]
+}

--- a/testdata/issues/issue170.wit.json.golden.wit
+++ b/testdata/issues/issue170.wit.json.golden.wit
@@ -1,0 +1,12 @@
+package issues:issue170;
+
+world w {
+	export example: interface {
+		resource r {
+			constructor(a: f64);
+			get-a: func() -> f64;
+			set-a: func(a: f64);
+			add: static func(r: r, a: f64) -> r;
+		}
+	}
+}

--- a/wit/bindgen/generator.go
+++ b/wit/bindgen/generator.go
@@ -544,10 +544,9 @@ func typeDefOwner(t *wit.TypeDef) wit.Ident {
 		id.Extension = owner.Name
 	case *wit.Interface:
 		id = owner.Package.Name
-		if owner.Name == nil {
-			id.Extension = "unknown"
-		} else {
-			id.Extension = *owner.Name
+		id.Extension = owner.InterfaceName()
+		if id.Extension == "" {
+			panic(fmt.Sprintf("BUG: unnamed interface owner of %#v", t))
 		}
 	}
 	return id

--- a/wit/resolve.go
+++ b/wit/resolve.go
@@ -8,9 +8,9 @@ import (
 	"strings"
 	"unsafe"
 
-	"github.com/coreos/go-semver/semver"
 	"github.com/bytecodealliance/wasm-tools-go/wit/iterate"
 	"github.com/bytecodealliance/wasm-tools-go/wit/ordered"
+	"github.com/coreos/go-semver/semver"
 )
 
 // Resolve represents a fully resolved set of WIT ([WebAssembly Interface Type])
@@ -125,6 +125,44 @@ type Interface struct {
 	Package   *Package  // the Package this Interface belongs to
 	Stability Stability // WIT @since or @unstable (nil if unknown)
 	Docs      Docs
+}
+
+// InterfaceName performs a best-effort guess at the [Interface] name,
+// which may be found by scanning its parent package.
+func (i *Interface) InterfaceName() string {
+	if i.Name != nil {
+		return *i.Name
+	}
+	if i.Package == nil {
+		return ""
+	}
+	var done bool
+	var name string
+	scanWorldItem := func(key string, item WorldItem) bool {
+		if ref, ok := item.(*InterfaceRef); ok && ref.Interface == i && key != "" {
+			name = key
+			done = true
+		}
+		return !done
+	}
+	i.Package.Worlds.All()(func(_ string, w *World) bool {
+		w.Imports.All()(scanWorldItem)
+		if !done {
+			w.Exports.All()(scanWorldItem)
+		}
+		return !done
+	})
+	if done {
+		return name
+	}
+	i.Package.Interfaces.All()(func(key string, face *Interface) bool {
+		if face == i && key != "" {
+			name = key
+			done = true
+		}
+		return !done
+	})
+	return name
 }
 
 // AllFunctions returns a [sequence] that yields each [Function] in an [Interface].

--- a/wit/testdata_test.go
+++ b/wit/testdata_test.go
@@ -361,10 +361,9 @@ func TestInterfaceNameIsNotNil(t *testing.T) {
 					name += "#" + *face.Name
 				}
 				t.Run(name, func(t *testing.T) {
-					// TODO: fix this, since anonymous imported interfaces have nil Name field
-					// if face.Name == nil {
-					// 	t.Error("Name is nil")
-					// }
+					if face.InterfaceName() == "" {
+						t.Error("empty or nil name")
+					}
 				})
 			}
 		})
@@ -421,8 +420,8 @@ func TestTypeDefRootOwnerNamesNotNil(t *testing.T) {
 							t.Error("Owner.Name is empty")
 						}
 					case *Interface:
-						if owner.Name == nil {
-							t.Error("Owner.Name is nil")
+						if owner.InterfaceName() == "" {
+							t.Error("Owner.Name is nil or Owner.InterfaceName() is empty")
 						}
 					}
 				})

